File: 41-50/Q44.cpp
//Q44.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int i = 0, j = 0;
        int star = -1, match = 0;

        while (i < s.size()) {
            if (j < p.size() && (p[j] == s[i] || p[j] == '?')) {
                i++; j++;                     // direct match
            }
            else if (j < p.size() && p[j] == '*') {
                star = j++;                  // record star position
                match = i;                   // match start
            }
            else if (star != -1) {           // backtrack to last star
                j = star + 1;
                i = ++match;
            }
            else return false;
        }

        while (j < p.size() && p[j] == '*') j++;  // trailing *

        return j == p.size();
    }
};

File: 41-50/Q44.py
#Q44.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        i = j = 0
        star = -1
        match = 0

        while i < len(s):
            if j < len(p) and (p[j] == s[i] or p[j] == '?'):
                i += 1
                j += 1
            elif j < len(p) and p[j] == '*':
                star = j
                j += 1
                match = i
            elif star != -1:
                j = star + 1
                match += 1
                i = match
            else:
                return False

        while j < len(p) and p[j] == '*':
            j += 1

        return j == len(p)
